File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/logic/simple/NakedSingles.kt
package sudokusolver.kotlin.logic.simple

import sudokusolver.kotlin.Board
import sudokusolver.kotlin.Cell
import sudokusolver.kotlin.SetValue
import sudokusolver.kotlin.UnsolvedCell

/*
 * http://www.sudokuwiki.org/Getting_Started
 *
 * If an unsolved cell has exactly one candidate, then the candidate must be placed in that cell.
 *
 * For each unsolved cell
 *   If the cell has one candidate
 *     Set the candidate as the value for the cell
 */
fun nakedSingles(board: Board<Cell>): List<SetValue> =
    board.cells
        .filterIsInstance<UnsolvedCell>()
        .filter { it.candidates.size == 1 }
        .map { SetValue(it, it.candidates.single()) }

File: sudokusolver_scala/src/main/scala/sudokusolver/scala/logic/simple/NakedSingles.scala
package sudokusolver.scala.logic.simple

import sudokusolver.scala.{Board, Cell, SetValue, UnsolvedCell}

/*
 * http://www.sudokuwiki.org/Getting_Started
 *
 * If an unsolved cell has exactly one candidate, then the candidate must be placed in that cell.
 */
def nakedSingles(board: Board[Cell]): Seq[SetValue] =
  board.cells.collect { case cell: UnsolvedCell if cell.candidates.size == 1 => SetValue(cell, cell.candidates.head) }